### PR TITLE
test: Tokenomics fork test

### LINKED ISSUE #939: Tokenomics launch: fork test

### PR DESCRIPTION
**Pull request by rackstar**
## Context

Milestone PR #1000! 🎉

Fork tests for Tokenomics RAMM upgrade. Includes swap tests and upgraded contract storage tests

**NOTE:**
This PR is behind PR #998 and PR #982. (contains both changes from those PRs)

## Changes proposed in this pull request

* deploy new Ramm contract
* upgrade existing contracts:
   *  LegacyGateway
   *  MCR
   *  Pool
   *  Cover
   * Assessment
   * LegacyPooledStaking
   * TokenController
   * IndividualClaims
   * YieldTokenIncidents
* add Ramm swap tests
   * swap ETH for NXM
   * swap NXM for ETH
* add upgraded contract value check:
   * Pool
   * MCR  
* add upgraded contract storage tests:
   * Pool
   * PriceFeedOracle
   * LegacyGateway

## Test plan

* added swap tests
* added upgraded contract storage tests

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [x] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

Closes #939

### CODE REVIEW

**Comment by rackstar** [comment]
Please double-check the newly added and upgraded contracts are correct based on the diff:

[release-candidate...feature/tokenomics ](https://github.com/NexusMutual/smart-contracts/compare/release-candidate...feature/tokenomics)

**Comment by rackstar** [comment]
These contracts are upgraded but not quite sure if its worth testing / how to test their storage:

[Assessment.sol](https://github.com/NexusMutual/smart-contracts/blob/feature/tokenomics/contracts/modules/assessment/Assessment.sol)
```javascript
  it('Compares storage of upgrade Assessment contract', async function () {
    // Assessment
    // stakeOf (address => Stake)
    // votesOf (address => Vote[])
    // hasAlreadyVotedOn (address => mapping(uint => bool))
    // assessments (Assessment[])
  });
```
[IndividualClaim.sol](https://github.com/NexusMutual/smart-contracts/blob/feature/tokenomics/contracts/modules/assessment/IndividualClaims.sol)
```javascript
  it('Compares storage of upgrade IndividualClaims contract', async function () {
    // IndividualClaims
    // lastClaimSubmissionOnCover (uint => ClaimSubmission)
    // claims (Claim[])
  });
```
[YieldTokenIncidents.sol](https://github.com/NexusMutual/smart-contracts/blob/feature/tokenomics/contracts/modules/assessment/YieldTokenIncidents.sol)
```javascript
  it('Compares storage of upgrade YieldTokenIncidents contract', async function () {
    // YieldTokenIncidents
    // incidents (Incident[])
  });
```
[TokenController.sol](https://github.com/NexusMutual/smart-contracts/blob/feature/tokenomics/contracts/modules/token/TokenController.sol)
```javascript
  it('Compares storage of upgrade TokenController contract', async function () {
    // TokenController
    // coverInfo (coverId => CoverInfo)
    // stakingPoolNXMBalances (pool id => { rewards, deposits })
  });
```
[Cover.sol](https://github.com/NexusMutual/smart-contracts/blob/feature/tokenomics/contracts/modules/cover/Cover.sol)
```javascript
  it('Compares storage of upgrade Cover contract', async function () {
    // Cover
    // coverSegmentAllocations (cover id => segment id => pool allocations[])
    // allowedPools (product id => allowed pool ids)
    // activeCover (assetId => { lastBucketUpdateId, totalActiveCoverInAsset })
    // productNames (productId => product name)
    // productTypeNames (productTypeId => productType name)
  });
```

**Comment by rackstar** [comment]
Tests are failing because PR https://github.com/NexusMutual/smart-contracts/pull/998 tests are failing. When those are fixed, i'll rebase and tests should pass

File: test/fork/product-type-update.js [hunk 1]
```diff
@@ -93,7 +93,6 @@ describe('Update product types', function () {
     );
     this.coverViewer = await ethers.getContractAt('CoverViewer', V2Addresses.CoverViewer);
     this.assessment = await ethers.getContractAt('Assessment', await this.master.getLatestAddress(toUtf8Bytes('AS')));
-    this.assessment = await ethers.getContractAt('Assessment', await this.master.getLatestAddress(toUtf8Bytes('AS')));
```
Comment (rackstar) [inline test/fork/product-type-update.js:96]: duplicate code

File: test/fork/tokenomics.js [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+    this.contractData.individualClaims.after.claim1 = await this.individualClaims.claims(1);
+    this.contractData.individualClaims.after.claimSubm1 = await this.individualClaims.lastClaimSubmissionOnCover(1);
+
+    Object.entries(this.contractData.individualClaims.before).forEach(([key, value]) => {
+      expect(this.contractData.individualClaims.after[key], assertionErrorMsg(key)).to.be.deep.equal(value);
+    });
+  });
+
+  it('Compares storage of upgrade YieldTokenIncidents contract', async function () {
+    await expect(this.yieldTokenIncidents.incidents(0)).to.be.reverted; // empty storage
```
Comment (rackstar) [inline test/fork/tokenomics.js:466]: this always throws an error and from the trace it returns 0x0 so I'm assuming its empty?
```js
await this.yieldTokenIncidents.incidents(0);
```
incidents storage:
```solidity
Incident[] public override incidents;
```

File: test/fork/tokenomics.js [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+    this.contractData.cover.after.productNames1 = await this.cover.productNames(1);
+    this.contractData.cover.after.productTypeNames1 = await this.cover.productTypeNames(1);
+
+    Object.entries(this.contractData.cover.after).forEach(([key, value]) => {
+      expect(this.contractData.cover.after[key], assertionErrorMsg(key)).to.be.deep.equal(value);
+    });
+
+    // Empty storage
+    await expect(this.cover.coverSegmentAllocations(1, 1, 1)).to.be.reverted;
+    await expect(this.cover.allowedPools(1, 1)).to.be.reverted;
```
Comment (rackstar) [inline test/fork/tokenomics.js:489]: These 2 also always throws an error and returns 0x0 on the trace.
Storage:
```solidity
mapping(uint => mapping(uint => PoolAllocation[])) public coverSegmentAllocations;
mapping(uint => uint[]) public allowedPools;
```

**Comment by roxdanila** [comment]
TODO: test the emergency pause for the ramm

**Review by roxdanila — APPROVED** [review]

**Pull request merged**